community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledEquivalenceUtils.java

```diff
@@ -0,0 +1,563 @@
+        {
+            return hashCode( (short[]) element );
+        }
+        else if ( element instanceof int[] )
+        {
+            return hashCode( (int[]) element );
+        }
+        else if ( element instanceof long[] )
+        {
+            return hashCode( (long[]) element );
+        }
+        else if ( element instanceof char[] )
+        {
+            return hashCode( (char[]) element );
+        }
+        else if ( element instanceof float[] )
+        {
+            return hashCode( (float[]) element );
+        }
+        else if ( element instanceof double[] )
+        {
+            return hashCode( (double[]) element );
+        }
+        else if ( element instanceof boolean[] )
+        {
+            return hashCode( (boolean[]) element );
+        }
+        else
+        {
+            return element.hashCode();
+        }
+    }
+
+    /**
+     * Calculate hash code of a map
+     * @param map the element to calculate hash code for
+     * @return the hash code of the given map
+     */
+    public static int hashCode( Map<String,Object> map )
+    {
+        int h = 0;
+        for ( Map.Entry<String,Object> next : map.entrySet() )
+        {
+            String k = next.getKey();
+            Object v = next.getValue();
+            h += (k == null ? 0 : k.hashCode()) ^ (v == null ? 0 : hashCode( v ));
+        }
+        return h;
+    }
+
+    /**
+     * Calculate hash code of a long value
+     * @param value the value to compute hash code for
+     * @return the hash code of the given value
+     */
+    public static int hashCode( long value )
+    {
+        return Long.hashCode( value );
+    }
+
+    /**
+     * Calculate hash code of a boolean value
+     * @param value the value to compute hash code for
+     * @return the hash code of the given value
+     */
+    public static int hashCode( boolean value )
+    {
+        return Boolean.hashCode( value );
+    }
+
+    /**
+     * Calculate hash code of a char value
+     * @param value the value to compute hash code for
+     * @return the hash code of the given value
+     */
+    public static int hashCode( char value )
+    {
+        return Character.hashCode( value );
+    }
+
+    /**
+     * Calculate hash code of a char[] value
+     * @param array the value to compute hash code for
+     * @return the hash code of the given value
+     */
+    public static int hashCode( char[] array )
+    {
+        int len = array.length;
+        switch ( len )
+        {
+        case 0:
+            return 42;
+        case 1:
+            return hashCode( array[0] );
+        case 2:
+            return 31 * hashCode( array[0] ) + hashCode( array[1] );
+        case 3:
+            return (31 * hashCode( array[0] ) + hashCode( array[1] )) * 31  + hashCode( array[2] );
+        default:
+            return len * (31 * hashCode( array[0] ) + hashCode( array[len / 2] ) * 31 + hashCode( array[len - 1] ));
```

Comment:
I see now that this comes from the `org.neo4j.cypher.internal.compiler.v3_2.commands.predicates.Equivalent` class